rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,228 @@
+
+For the purpose of automating deployments:
+
+- We only need to add contract tests when it's hard to be confident about an API change. If an API endpoint only has a single "consumer" app, then it's easy to gain confidence by manually testing the pair still work together.
+
+- We only need to add contract tests for the APIs we consume internally. Some of our APIs may be public, but this is a separate, product-level commitment. Our strategy for testing public APIs is out-of-scope for this RFC.
+
+With this in mind, an API app has "enough" of these tests when:
+
+- Each endpoint with multiple consumers has at least one contract test.
+
+> Example: [Asset Manager](https://github.com/alphagov/asset-manager) is missing contract tests because multiple apps use it to create [[1](https://github.com/alphagov/content-publisher/blob/0c757447ca2aad3621f11c99fe8307a718ade186/app/services/preview_asset_service.rb#L29)] [[2](https://github.com/alphagov/specialist-publisher/blob/8debeb2f0147142c87f22308d57fe4f6dd0c1297/app/models/attachment.rb#L49)] assets, among other APIs. Conversely, no [Support API](https://github.com/alphagov/support-api) endpoints are used by more than one GOV.UK app [[1](https://github.com/alphagov/support-api/tree/75540575ae11095bb8d4736490abdc9927049aaf#technical-documentation)].
+
+#### Check: Smoke tests pass (i.e. the app runs in a production environment)
+
+For the purpose of automating deployments:
+
+- We need to test that an app can respond to a request for each of the machine classes it runs on [[1](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/content-store/config/deploy.rb#L3)] [[2](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/whitehall/config/deploy.rb#L13)]. This is purely to prove that an instance of the app is running [[1](https://github.com/alphagov/email-alert-api/pull/1346)], and may be covered by other tests.
+
+- We need to test that an app can connect to any systems it uses for data storage and retrieval, such as a database. This usually has production-specific config [[1](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/database.yml#L22)] [[2](https://github.com/alphagov/content-store/blob/f4b61d4f2756341b6418085d878ec83b09d70e9c/config/mongoid.yml#L23)] [[3](https://github.com/alphagov/content-publisher/blob/944fe51b4cb8205348e155b965b8bad57adcf4fe/config/storage.yml#L5)], which could be faulty.
+
+- We do not need to test connectivity to other GOV.UK apps, since this should be handled by [our internal API adapters](https://github.com/alphagov/gds-api-adapters/blob/1ceba690edb37295257328d953874224833fcd5e/lib/gds_api.rb) with no configuration inside the repo. An exception to this is [authentication with Signon](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/initializers/gds_sso.rb).
```

Comment:
What are the testing requirements for important libraries such as gds-api-adapters? Would we catch, say for example, a bug which caused wrong URLs to be generated for certain API endpoints under certain conditions?

Reply:
For gds-api-adapters, this should be covered by the previous section: we will require a minimum level of testing for this library, since the way the code in it behaves is one of the main differences between test and production for apps that use the adapters (which is basically all of them!).

For other libraries, we have no additional test requirements. This is because the code in our apps that makes use of them doesn't vary between test and production, so nothing changes* between manual vs. automatic deployments for those apps. Definitely part of a broader testing strategy, though.

*An interesting example is [the Slimmer library](https://github.com/alphagov/slimmer/blob/64482b28572a7d038827de214832317e75f0c1ae/lib/slimmer/test.rb), which is the thing we use to shove the GOV.UK header/footer on to pages rendered by our frontend apps (by calling a rendering API). A bug in this library could cause havoc, but it has no code footprint in our apps. So, while the gem itself (still) needs to be really thoroughly tested, it has no code in our apps that could be broken in a PR (that I've been able to find).

Does that address your concern here?